Libraries/LibPthread/pthread.cpp

```diff
@@ -157,70 +167,233 @@ int pthread_sigmask(int how, const sigset_t* set, sigset_t* old_set)
     return 0;
 }
 
-int pthread_mutex_init(pthread_mutex_t* mutex, const pthread_mutexattr_t* attributes)
+#define PTHREAD_FUTEX_UNLOCKED 0u
+#define PTHREAD_FUTEX_LOCKED 1u
+#define PTHREAD_FUTEX_LOCKED_WITH_WAITERS 2u
+
+static int futex_lock(uint32_t& futex_addr, uint32_t futex_flags, const timespec* abstime = nullptr, bool no_wait = false)
 {
-    mutex->lock = 0;
-    mutex->owner = 0;
-    mutex->level = 0;
-    mutex->type = attributes ? attributes->type : PTHREAD_MUTEX_NORMAL;
-    return 0;
+    ASSERT(!(futex_flags & ~(FUTEX_PRIVATE_FLAG | FUTEX_CLOCK_REALTIME)));
+    for (;;) {
+        uint32_t expected = PTHREAD_FUTEX_UNLOCKED;
+        if (AK::atomic_compare_exchange_strong(&futex_addr, expected, PTHREAD_FUTEX_LOCKED, AK::MemoryOrder::memory_order_acquire)) {
+            atomic_thread_fence(AK::MemoryOrder::memory_order_release);
+#ifdef PTHREAD_DEBUG
+            dbg() << "futex_lock @ " << &futex_addr << ": acquired @ " << &futex_addr;
+#endif
+            return 0;
+        }
+        if (no_wait)
+            return EBUSY;
+        if (expected == PTHREAD_FUTEX_LOCKED_WITH_WAITERS || AK::atomic_compare_exchange_strong(&futex_addr, expected, PTHREAD_FUTEX_LOCKED_WITH_WAITERS, AK::MemoryOrder::memory_order_acquire)) {
+#ifdef PTHREAD_DEBUG
+            dbg() << "futex_lock @ " << &futex_addr << ": waiting";
+#endif
+            // NOTE: FUTEX_WAIT takes a relative timeout, so use FUTEX_WAIT_BITSET instead!
+            int saved_errno = errno;
+            int rc = futex(&futex_addr, FUTEX_WAIT_BITSET | futex_flags, PTHREAD_FUTEX_LOCKED_WITH_WAITERS, abstime, nullptr, FUTEX_BITSET_MATCH_ANY);
+            if (rc < 0) {
+                int err = errno;
+                errno = saved_errno;
+                if (err == EAGAIN) {
+#ifdef PTHREAD_DEBUG
+                    dbg() << "futex_lock @ " << &futex_addr << ": waiting aborted, try locking again";
+#endif
+                    continue;
+                }
+#ifdef PTHREAD_DEBUG
+                dbg() << "futex_lock @ " << &futex_addr << ": wait error " << err;
+#endif
+                return err;
+            }
+            atomic_thread_fence(AK::MemoryOrder::memory_order_release);
+#ifdef PTHREAD_DEBUG
+            dbg() << "futex_lock @ " << &futex_addr << ": acquired after waiting";
+#endif
+            return 0;
```

Comment:
After we wake from sleeping on a futex, we need to loop again & try to take the lock, otherwise we leave it unlocked

Reply:
Dang you're right. BUT this is the spot to leverage `FUTEX_WAKE_OP` instead so this looping and taking the lock is not necessary ;-)

Reply:
I don't think it's true though. You still need the thread to "see" the stuff the other thread has written, so it needs to perform an acquire op *after* waking from sleep. It can be just a fence though.

But there's another tricky thing here. The unlock/wake path has to actually change the value (from `locked_with_waiters` to `unlocked`, presumably) for the lock/wait path not to miss wakeups.

P.S. I assume you have read "futexes are tricky", right?

Reply:
Yeah I've read that. I think that at least the kernel support is in good shape (although largely untested at this point).

Thinking about it some more I don't think `FUTEX_WAKE_OP` will work in this particular instance, either.